### DISCUSSION TITLE
Discrepancy between dipy.gibbs.gibbs_removal and reisert/unring/

### DISCUSSION THREAD

**Discussion by tashrifbillah** [dipy/dipy#2054] [Preprocessing]
Hi @RafaelNH 

## Description

Has the Gibbs unringing algorithm been reproduced from https://bitbucket.org/reisert/unring/src/ ? If the answer is yes, I am observing large difference at voxel level (+-100) between the outputs of `reisert/unring/src/fsl/unring.a64` and `dipy.gibbs.gibbs_removal` for a 3D volume.

## Way to reproduce

I extracted a baseline image from a DWI. Consider the name of the image `b0.nii.gz`

* reisert unring

> unring/fsl/unring.a64 b0.nii.gz reisert.nii.gz

* dipy unring

```python
import nibabel as nib
from dipy.denoise.gibbs import gibbs_removal

img= nib.load('b0.nii.gz')
unringed_volume= gibbs_removal(img.get_data())

nib.Nifti1Image(unringed_volume, affine= img.affine, header= img.header).to_filename('dipy.nii.gz')

```

You can compare `dipy.nii.gz` and `reisert.nii.gz` using your favourite visualizer. I used fsleyes:

> fsleyes dipy.nii.gz reisert.nii.gz

You should see at each voxel, the values differ by +-100.

You can also check the difference in IPython

```python
diff= abs(nib.load('dipy.nii.gz).get_data()- nib.load('reisert.nii.gz'))
diff.mean()
diff.std()
# etc
```

Actually, the discrepancy is observed right at the documentation of reisert and dipy:

```bash
unring.a64 -h

unring - tool for removal of the Gibbs ringing artefact
Usage: unring <input> <output> [options]
 Options: -d     slice direction, in the perpendicular plane unringing is performed. Could be either 1,2 or 3  (default 3)
          -nsh   discretization of subpixel spaceing (default 20)
          -minW  left border of window used for TV computation (default 1)
          -maxW  right border of window used for TV computation (default 3)
```

https://github.com/dipy/dipy/blob/f2e5ecef63b6fc609bda3b6f581a212ae1ad8a35/dipy/denoise/gibbs.py#L223
```python
    vol : ndarray ([X, Y]), ([X, Y, Z]) or ([X, Y, Z, g])
        Matrix containing one volume (3D) or multiple (4D) volumes of images.
    slice_axis : int (0, 1, or 2)
        Data axis corresponding to the number of acquired slices.
        Default is set to the third axis.
    n_points : int, optional
        Number of neighbour points to access local TV (see note).
        Default is set to 3.

```

What would be the equivalent of `nsh` in dipy? Have `minW` and `maxW` been combined to `n_points` ? Do these simplification yield the final difference I observed?

- [x] dipy version (run `python -c "import dipy; print(dipy.__version__)"`)

> 0.16.0

- [x] dependency version (numpy, scipy, nibabel, h5py, cvxpy, fury)
    * import numpy; print("NumPy", numpy.__version__)
    * import scipy; print("SciPy", scipy.__version__)
    * import nibabel; print("Nibabel", nibabel.__version__)
    * import h5py; print("H5py", h5py.__version__)

```
NumPy 1.16.2
SciPy 1.2.1
Nibabel 2.3.0
H5py 2.9.0
```

**Comment by tashrifbillah** [comment]
cc: @sbouix

**Comment by ShreyasFadnavis** [comment]
Hi @tashrifbillah ! Thank you for reporting this 👍 
@RafaelNH will get back to you on this soon...

**Comment by RafaelNH** [comment]
Hi @tashrifbillah ! As I've mentioned on Gibbs tutorial, this algorithm is an adaption of Kellner's method, so it might produce a slightly different outcome. Full details of the implemented algorithm can be found in chapter 3 of my PhD thesis (see the references on Gibb's tutorial).

**Comment by tashrifbillah** [comment]
Hi @RafaelNH , can you point me to the Gibb's tutorial?

**Comment by skoudoro** [comment]
Hi @tashrifbillah,

Here you are for the tutorial: https://dipy.org/documentation/1.1.1./examples_built/denoise_gibbs/#example-denoise-gibbs

**Comment by tashrifbillah** [comment]
Hi @RafaelNH , can you point us to your PhD thesis, more specifically Chapter 3?
I did follow the [link](https://digital.lib.washington.edu/researchworks/handle/1773/33311) from docstring but couldn't find the above.

**Comment by tashrifbillah** [comment]
Ping @RafaelNH ...

**Comment by skoudoro** [comment]
Hi @tashrifbillah,

What is the status of this issue? Can we close it?

As @RafaelNH recommend you, you can find his PhD thesis in the following link: https://www.repository.cam.ac.uk/handle/1810/281993 or download it [here](https://www.repository.cam.ac.uk/bitstream/handle/1810/281993/2017_RafaelNH_PhD.pdf?sequence=1&isAllowed=y)

**Comment by skoudoro** [comment]
closing this issue, feel free to reopen it at any moment @tashrifbillah.
